buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesGraphTask.java

```diff
@@ -0,0 +1,154 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.gradle.api.GradleException;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.DependencySet;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.TaskAction;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A task to generate a dependency graph of our runtime dependencies and push that via
+ * an API call to a given endpoint of a SCA tool/service.
+ * The graph is built according to the specification in https://github.com/snyk/dep-graph#depgraphdata
+ *
+ * Due to the nature of our dependency resolution in gradle, we are abusing the aforementioned graph definition as
+ * the graph we construct has a single root ( the subproject ) and all dependencies are children of that root,
+ * irrespective of if they are direct dependencies or transitive ones ( that should be children of other children ).
+ * Although we end up lacking some contextual information, this allows us to scan and monitor only the dependencies
+ * that are bundled and used in runtime.
+ */
+public class DependenciesGraphTask extends DefaultTask {
+
+    private Configuration runtimeConfiguration;
+    private Configuration compileOnlyConfiguration;
+
+    @InputFiles
+    public Configuration getRuntimeConfiguration() {
+        return runtimeConfiguration;
+    }
+
+    @InputFiles
+    public Configuration getCompileOnlyConfiguration() {
+        return compileOnlyConfiguration;
+    }
+
+    public void setRuntimeConfiguration(Configuration runtimeConfiguration) {
+        this.runtimeConfiguration = runtimeConfiguration;
+    }
+
+    public void setCompileOnlyConfiguration(Configuration compileOnlyConfiguration) {
+        this.compileOnlyConfiguration = compileOnlyConfiguration;
+    }
+
+    @TaskAction
+    void generateDependenciesGraph() {
+
+        if (getProject().getGradle().getStartParameter().isOffline()) {
+            throw new GradleException("Must run in online mode in order to submit the dependency graph to the SCA service");
+        }
+        final String url = System.getenv("SCA_URL");
+        final String token = System.getenv("SCA_TOKEN");
+        if (null == url || null == token) {
```

Comment:
Can we make these properties on the task, and then set them in the plugin?

Reply:
Sure thing. What's the benefit of that ? i.e. why do we prefer to do that in the plugin?

Reply:
The thing I dislike about this is that if the env vars are not set , this now fails with 
```
* What went wrong:
Some problems were found with the configuration of task ':rest-api-spec:dependenciesGraph' (type 'DependenciesGraphTask').
> No value has been specified for property 'token'.
> No value has been specified for property 'url'.
```

instead of the Gradle exception with the proper message before. 

Is there any way to add a more user friendly message to the Task failing if an `@Input` is passed as null ? I guess I can also annotate those as `@Optional` and do the null check myself but it sounds like a bad solution as these are not optional at all :)

Reply:
> Sure thing. What's the benefit of that ? i.e. why do we prefer to do that in the plugin?

The general pattern is that plugins should "wire" up tasks. So reaching out to environment variables should be done in the "wiring" no in the implementation of the task.

> Is there any way to add a more user friendly message to the Task failing if an @Input is passed as null ? I guess I can also annotate those as @Optional and do the null check myself but it sounds like a bad solution as these are not optional at all :)

They are optional in the sense that they aren't required unless the task is actually going to execute. So what we instead need to do is add a task graph listener that explodes if a) that task is included in the graph for execution and b) the environment variables are missing ([example](https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportPlugin.java#L54-L65)).

Reply:
Thanks for explaining Mark !